spyder/plugins/editor/widgets/tests/test_introspection.py

```diff
@@ -353,6 +368,177 @@ def test_completions(lsp_codeeditor, qtbot):
     code_editor.toggle_automatic_completions(True)
+    code_editor.toggle_code_snippets(True)
+
+
+@pytest.mark.slow
+@pytest.mark.first
+@pytest.mark.skipif(not rtree_available or PY2,
+                    reason='Only works if libspatialindex is installed')
+def test_code_snippets(lsp_codeeditor, qtbot):
+    assert rtree_available
+    code_editor, lsp = lsp_codeeditor
+    completion = code_editor.completion_widget
+    snippets = code_editor.editor_extensions.get('SnippetsExtension')
+
+    CONF.set('lsp-server', 'code_snippets', True)
+    lsp.update_configuration()
+
+    code_editor.toggle_automatic_completions(False)
+    code_editor.toggle_code_snippets(True)
+    # Set cursor to start
+    code_editor.go_to_line(1)
+
+    text = """
+    def test_func(xlonger, y1, some_z):
+        pass
+    """
+    text = textwrap.dedent(text)
+
+    # qtbot.keyClicks(code_editor, 'import json')
+    code_editor.insert_text(text)
+    with qtbot.waitSignal(code_editor.lsp_response_signal, timeout=30000):
+        code_editor.document_did_change()
+
+    qtbot.keyPress(code_editor, Qt.Key_Enter, delay=300)
+    qtbot.keyPress(code_editor, Qt.Key_Enter, delay=300)
+    qtbot.keyClicks(code_editor, 'test_')
+    with qtbot.waitSignal(completion.sig_show_completions,
+                          timeout=10000) as sig:
+        qtbot.keyPress(code_editor, Qt.Key_Tab)
+
+    print({x['label'] for x in sig.args[0]})
+    assert 'test_func(xlonger, y1, some_z)' in {
+        x['label'] for x in sig.args[0]}
+
+    expected_insert = 'test_func(${1:xlonger}, ${2:y1}, ${3:some_z})$0'
+    insert = sig.args[0][0]
+    assert expected_insert == insert['insertText']
+
+    assert snippets.is_snippet_active
+    assert code_editor.has_selected_text()
+
+    # Rotate through snippet regions
+    cursor = code_editor.textCursor()
+    arg1 = cursor.selectedText()
+    assert 'xlonger' == arg1
+    assert snippets.active_snippet == 1
+
+    qtbot.keyPress(code_editor, Qt.Key_Tab)
+    cursor = code_editor.textCursor()
+    arg2 = cursor.selectedText()
+    assert 'y1' == arg2
+    assert snippets.active_snippet == 2
+
+    qtbot.keyPress(code_editor, Qt.Key_Tab)
+    cursor = code_editor.textCursor()
+    arg2 = cursor.selectedText()
+    assert 'some_z' == arg2
+    assert snippets.active_snippet == 3
+
+    qtbot.keyPress(code_editor, Qt.Key_Tab)
+    assert snippets.active_snippet == 0
+
+    qtbot.keyPress(code_editor, Qt.Key_Tab)
+    assert snippets.active_snippet == 1
+    assert code_editor.has_selected_text()
+
+    # Replace selection
+    qtbot.keyClicks(code_editor, 'arg1')
+    qtbot.wait(5000)
+
+    assert len(snippets.snippets_map) == 4
+    assert snippets.active_snippet == 1
+
+    qtbot.keyPress(code_editor, Qt.Key_Right, delay=300)
+    qtbot.keyPress(code_editor, Qt.Key_Tab)
+    qtbot.keyPress(code_editor, Qt.Key_Tab)
+    qtbot.keyPress(code_editor, Qt.Key_Tab)
+    qtbot.keyPress(code_editor, Qt.Key_Tab)
```

Comment:
Let's use a `for` here to generate four `Qt.Key_Tab` key presses instead of repeating the same line several times.